src/uct/ib/base/ib_md.c

```diff
@@ -1694,31 +1705,74 @@ ucs_status_t uct_ib_md_open_common(uct_ib_md_t *md,
 
     return UCS_OK;
 
-err_dealloc_pd:
-    ibv_dealloc_pd(md->pd);
 err_cleanup_device:
     uct_ib_device_cleanup(&md->dev);
 err_release_stats:
     UCS_STATS_NODE_FREE(md->stats);
+err_device_config_release:
+    uct_ib_md_release_device_config(md);
 err:
     return status;
 }
 
-void uct_ib_md_close(uct_md_h uct_md)
+void uct_ib_md_close_common(uct_ib_md_t *md)
 {
-    uct_ib_md_t *md = ucs_derived_of(uct_md, uct_ib_md_t);
-
-    /* Must be done before md->ops->cleanup, since it can call functions from
-     * md->ops */
     uct_ib_md_release_reg_method(md);
-    md->ops->cleanup(md);
-    uct_ib_md_release_device_config(md);
-    uct_ib_device_cleanup_ah_cached(&md->dev);
-    ibv_dealloc_pd(md->pd);
     uct_ib_device_cleanup(&md->dev);
-    ibv_close_device(md->dev.ibv_context);
     UCS_STATS_NODE_FREE(md->stats);
+    uct_ib_md_release_device_config(md);
+}
+
+uct_ib_md_t* uct_ib_md_alloc(size_t size, const char *name,
+                             struct ibv_context *ctx)
+{
+    uct_ib_md_t *md;
+
+    md = ucs_calloc(1, size, name);
+    if (md == NULL) {
+        ucs_error("failed to allocate memory for md");
+        goto err;
+    }
+
+    md->dev.ibv_context = ctx;
+    md->pd = ibv_alloc_pd(md->dev.ibv_context);
+    if (md->pd == NULL) {
+        ucs_error("ibv_alloc_pd() failed: %m");
+        goto err_md_free;
+    }
+
+    return md;
+
+err_md_free:
     ucs_free(md);
+err:
+    return NULL;
+}
+
+void uct_ib_md_free(uct_ib_md_t *md)
+{
+    int ret;
+
+    ret = ibv_dealloc_pd(md->pd);
+    /* Do not print a warning if PD deallocation failed with EINVAL, because
+     * it fails from time to time on BF/ARM (TODO: investigate) */
+    if ((ret != 0) && (errno != EINVAL)) {
+        ucs_warn("ibv_dealloc_pd() failed: %m");
+    }
+
+    ret = ibv_close_device(md->dev.ibv_context);
```

Comment:
but uct_ib_md_alloc did not open a device

Reply:
but it sets device (i.e. moves ownershhip) which was allocated by some upper layer - Verbs/MLX5_DevX/MLX5_DV md_open.
otherwise - we need to duplicate `ibv_close_device` in all MD close functions

Reply:
> otherwise - we need to duplicate ibv_close_device in all MD close functions
IMO we should - to keep consistency